File: src/learning_tf2_py/learning_tf2_py/turtle_tf2_broadcaster.py
import math

from geometry_msgs.msg import TransformStamped
import numpy as np
import rclpy
from rclpy.node import Node
from tf2_ros import TransformBroadcaster
from turtlesim.msg import Pose


# 欧拉角 → 四元数（跟上一个教程一样）
def quaternion_from_euler(ai, aj, ak):
    ai /= 2.0
    aj /= 2.0
    ak /= 2.0
    ci = math.cos(ai)
    si = math.sin(ai)
    cj = math.cos(aj)
    sj = math.sin(aj)
    ck = math.cos(ak)
    sk = math.sin(ak)
    cc = ci * ck
    cs = ci * sk
    sc = si * ck
    ss = si * sk
    q = np.empty((4,))
    q[0] = cj * sc - sj * cs
    q[1] = cj * ss + sj * cc
    q[2] = cj * cs - sj * sc
    q[3] = cj * cc + sj * ss
    return q


class FramePublisher(Node):

    def __init__(self):
        super().__init__('turtle_tf2_frame_publisher')

        # ★ 1. 声明一个参数：海龟的名字（turtle1 或 turtle2）
        self.turtlename = self.declare_parameter(
            'turtlename', 'turtle'
        ).get_parameter_value().string_value

        # ★ 2. 创建动态变换广播器（不是静态！）
        self.tf_broadcaster = TransformBroadcaster(self)

        # ★ 3. 订阅 /turtleX/pose 话题
        #     每次收到位置数据就调用 handle_turtle_pose
        self.subscription = self.create_subscription(
            Pose,
            f'/{self.turtlename}/pose',
            self.handle_turtle_pose,
            1)
        self.subscription  # 防止"未使用变量"警告

    def handle_turtle_pose(self, msg):
        # 构造 TransformStamped 消息
        t = TransformStamped()

        # 时间戳
        t.header.stamp = self.get_clock().now().to_msg()
        # 父坐标系 = world
        t.header.frame_id = 'world'
        # 子坐标系 = turtle1（或 turtle2）
        t.child_frame_id = self.turtlename

        # 位置：海龟在 2D 平面，z=0
        t.transform.translation.x = msg.x
        t.transform.translation.y = msg.y
        t.transform.translation.z = 0.0

        # 旋转：海龟只能绕 Z 轴转，所以 roll=0, pitch=0
        q = quaternion_from_euler(0, 0, msg.theta)
        t.transform.rotation.x = q[0]
        t.transform.rotation.y = q[1]
        t.transform.rotation.z = q[2]
        t.transform.rotation.w = q[3]

        # 发布到 TF2
        self.tf_broadcaster.sendTransform(t)


def main():
    rclpy.init()
    node = FramePublisher()
    try:
        rclpy.spin(node)
    except KeyboardInterrupt:
        pass
    rclpy.shutdown()

File: src/learning_tf2_cpp/src/turtle_tf2_broadcaster.cpp
#include <memory>

#include "geometry_msgs/msg/transform_stamped.hpp"
#include "geometry_msgs/msg/twist.hpp"
#include "rclcpp/rclcpp.hpp"
#include "tf2/LinearMath/Quaternion.h"
#include "tf2_ros/transform_broadcaster.h"
#include "turtlesim/msg/pose.hpp"

class FramePublisher : public rclcpp::Node
{
public:
  FramePublisher()
  : Node("turtle_tf2_broadcaster")
  {
    // 创建 TransformBroadcaster
    tf_broadcaster_ = std::make_unique<tf2_ros::TransformBroadcaster>(*this);

    // 声明一个参数 "turtlename"，默认值是 "turtle1"
    this->declare_parameter<std::string>("turtlename", "turtle1");

    // 获取参数
    std::string turtlename = this->get_parameter("turtlename").as_string();

    // 订阅对应乌龟的 pose 话题
    subscription_ = this->create_subscription<turtlesim::msg::Pose>(
      "/" + turtlename + "/pose", 10,
      [this, turtlename](const turtlesim::msg::Pose::SharedPtr msg) {
        handle_turtle_pose(msg, turtlename);
      });
  }

private:
  void handle_turtle_pose(
    const turtlesim::msg::Pose::SharedPtr msg,
    const std::string & turtlename)
  {
    // 构造 TransformStamped 消息
    geometry_msgs::msg::TransformStamped t;

    // 时间戳用现在
    t.header.stamp = this->get_clock()->now();
    t.header.frame_id = "world";
    t.child_frame_id = turtlename;  // 比如 "turtle1"

    // 乌龟的 x, y 位置
    t.transform.translation.x = msg->x;
    t.transform.translation.y = msg->y;
    t.transform.translation.z = 0.0;

    // 乌龟朝向：从 yaw（偏航角）转成四元数
    tf2::Quaternion q;
    q.setRPY(0, 0, msg->theta);  // roll=0, pitch=0, yaw=theta
    t.transform.rotation.x = q.x();
    t.transform.rotation.y = q.y();
    t.transform.rotation.z = q.z();
    t.transform.rotation.w = q.w();

    // 发送变换
    tf_broadcaster_->sendTransform(t);
  }

  std::unique_ptr<tf2_ros::TransformBroadcaster> tf_broadcaster_;
  rclcpp::Subscription<turtlesim::msg::Pose>::SharedPtr subscription_;
};

int main(int argc, char * argv[])
{
  rclcpp::init(argc, argv);
  rclcpp::spin(std::make_shared<FramePublisher>());
  rclcpp::shutdown();
  return 0;
}
